### PR TITLE
add `mock-property`

### PR DESCRIPTION
**Pull request by ljharb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/mock-property/mock-property-tests.ts
1. types/mock-property/index.d.ts
Consider running `npx dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7ee48dcd2520eafb4771eef1dfa3eb9d532dd096
</p>

File: types/mock-property/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+declare function mockProperty(
+    obj: object,
+    prop: PropertyKey,
+    options?: (
+        | { delete: true }
+        | (
+            { nonEnumerable?: boolean, delete?: false }
+            & (
+                | { nonWritable?: boolean, value?: unknown }
+                | { get?: (() => unknown) | undefined, set?: ((v: unknown) => void) | undefined }
```
Comment (jakebailey) [inline types/mock-property/index.d.ts:17]: These `unknown`s make me feel like this function could really be generic and more type safe, but, it's probably fine for now.

**Review by jakebailey — APPROVED** [review]
What specifically does it complain about? The bot will auto format post-merge anyway, so it's not a merge requirement; you can just leave it however, though if there's a bug in dprint that'd be helpful to report upstream.

And of course, I'll reiterate prior feedback that it's really not thrilling to be pushing types to DT when you have full power to publish types with your packages. DT must necessarily _also_ match the versioning of the upstream package, but that means that anything that would be "semver breaking" at the original package is also "semver breaking" on DT, so I really don't understand why it's any less breaking to push types over to DT besides making maintenance our problem.

**Comment by ljharb** [comment]
It’s mainly that if *something* has to violate semver, I’d rather it be a DT package (since TS already plays fast and loose with it) than my own.

I’m continuing to explore using jsdoc and emitted d.ts types; I’ve got a package using it, this PR unblocks another, and a (filed) bug with invalid emit blocks another, so maybe over time I’ll change my preferred approach.

**Comment by ljharb** [comment]
Ready to merge

**Comment by jakebailey** [comment]
> I’m continuing to explore using jsdoc and emitted d.ts types; I’ve got a package using it, this PR unblocks another, and a (filed) bug with invalid emit blocks another, so maybe over time I’ll change my preferred approach.

I'm not totally sure that will have the effect you want, based on your criteria; that'll have the same amount of "breakness" as switching to TS or publishing types directly. Realistically emitting d.ts files from JSDoc is _worse_ than just using TS in the first place, but it's certainly doable so long as you avoid pitfalls about typedefs and so on. But, it's still moving the types.

**Comment by ljharb** [comment]
You're right; it's just an experiment, but preferring DT may remain the better choice.

**Pull request merged**